pytorch_lightning/utilities/cli.py

```diff
@@ -212,27 +217,11 @@ def before_instantiate_classes(self) -> None:
 
     def instantiate_classes(self) -> None:
         """Instantiates the classes using settings from self.config"""
-        self.config_init = self.parser.instantiate_subclasses(self.config)
-        self.instantiate_datamodule()
-        self.instantiate_model()
+        self.config_init = self.parser.instantiate_classes(self.config)
+        self.datamodule = self.config_init.get('data')
+        self.model = self.config_init['model']
         self.instantiate_trainer()
 
-    def instantiate_datamodule(self) -> None:
-        """Instantiates the datamodule using self.config_init['data'] if given"""
-        if self.datamodule_class is None:
-            self.datamodule = None
-        elif self.subclass_mode_data:
-            self.datamodule = self.config_init['data']
-        else:
-            self.datamodule = self.datamodule_class(**self.config_init.get('data', {}))
-
-    def instantiate_model(self) -> None:
-        """Instantiates the model using self.config_init['model']"""
-        if self.subclass_mode_model:
-            self.model = self.config_init['model']
-        else:
-            self.model = self.model_class(**self.config_init.get('model', {}))
-
```

Comment:
Why are these gone?

We will need them in Flash to instantiate the datamodule from a `classmethod`

Reply:
It is necessary to remove them to make it possible to have links that are applied at instantiation time. Based on the links the order of instantiation is determined. There is a loop, the first class is instantiated followed by applying the links that make use of that class, go to the second class and repeat.

Regarding flash I would need to know more about what is the situation there. Anyway note that currently these methods are inconsistent. In subclass mode the class has already been instantiated, whereas in normal mode the init args are used to instantiate. Don't know how this is used in flash but it probably has this subclass thing broken. Depending on the exact need in flash we can see how to do it. But it is till best to remove these methods.

Reply:
In Flash, `DataModule`s can be created using different `classmethods`.

Some examples are `from_folders`, `from_csv`, `from_tensors`, ...

https://github.com/PyTorchLightning/lightning-flash/blob/a047330ba75486355378f22cbebfd053c3d63c08/flash_examples/finetuning/image_classification.py#L28

A `DataModule` can support multiple of these.

But actually, I was being dumb, we just need to support the following:


```python
import torch
from torch.utils.data import Dataset, DataLoader

import pytorch_lightning as pl
from pytorch_lightning.utilities.cli import LightningCLI


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class MyDataModule(pl.LightningDataModule):
    @classmethod
    def from_thing(cls):
        return cls()

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 20))


class MyModel(pl.LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


cli = LightningCLI(MyModel, MyDataModule.from_thing, trainer_defaults={'fast_dev_run': True})
```

So passing a function to create the datamodule.

So there's no need to subclass and override `instantiate_datamodule`.

Should be easy right? Just need to call `add_function_arguments` if the passed input is a function instead of a class

Reply:
Yes, could be. However, with `add_function_arguments` the instantiation links wouldn't work. Though I do think something simple could be done to make them work.

Reply:
Do you want to open a PR with the improvement?